compiler/src/dotty/tools/dotc/transform/init/Cache.scala

```diff
@@ -0,0 +1,132 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core.*
+import Contexts.*
+
+import ast.tpd
+import tpd.Tree
+
+class Cache[Config, Res]:
```

Comment:
General documentation would be very useful here. In particular:

1. A very general overview of the algorithm that is meant to use this cache.
2. An explanation of what Config and Res could be, and that the cache is logically a map from (Config,Tree) pairs to Res.

Reply:
3. A postcondition that the algorithm should satisfy. Something like `eval(current, tree) = current(tree)`.

Making `eval` take no parameters, so that it gets its real parameters by capturing them from the enclosing scope, makes it difficult to understand what the algorithm does and how it works. If the parameters to eval were explicit, the desired postcondition of the algorithm (in terms of the cache) would be easier to state and to understand.